server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/CassandraSessionModule.java

```diff
@@ -133,4 +156,60 @@ ClusterConfiguration provideClusterConfiguration(PropertiesProvider propertiesPr
                 .build();
         }
     }
+
+    @Provides
+    @Singleton
+    KeyspacesConfiguration provideKeyspacesConfiguration(PropertiesProvider propertiesProvider) throws ConfigurationException {
+        try {
+            return KeyspacesConfiguration.from(propertiesProvider.getConfiguration(CASSANDRA_FILE_NAME));
+        } catch (FileNotFoundException e) {
+            LOGGER.warn("Could not locate cassandra configuration file. Using default keyspaces configuration instead");
+            return KeyspacesConfiguration.builder().build();
+        }
+    }
+
+    @Provides
+    @Singleton
+    KeyspaceConfiguration provideMainKeyspaceConfiguration(KeyspacesConfiguration keyspacesConfiguration) {
+        return keyspacesConfiguration.mainKeyspaceConfiguration();
+    }
+
+    @Named(InjectionNames.CACHE)
+    @Provides
+    @Singleton
+    KeyspaceConfiguration provideCacheKeyspaceConfiguration(KeyspacesConfiguration keyspacesConfiguration) {
+        return keyspacesConfiguration.cacheKeyspaceConfiguration();
+    }
+
+    private static class MainSessionWithInitializedTablesFactory extends SessionWithInitializedTablesFactory {
+        @Inject
+        public MainSessionWithInitializedTablesFactory(KeyspaceConfiguration keyspaceConfiguration,
+                                                       InitializedCluster cluster,
+                                                       CassandraModule module) {
+            super(keyspaceConfiguration, cluster.cluster, module);
+        }
+    }
+
+    private static class CacheSessionWithInitializedTablesFactory extends SessionWithInitializedTablesFactory {
```

Comment:
you don't need types for that, you can reuse CACHE constant to annotate SessionWithInitializedTablesFactory

Reply:
You can annotate the provided value but not the provider, and even less the providers constructor arguments.

Reply:
I don't understand what you are talking about.

Can't you write:

```java
@Named(InjectionNames.CACHE)
@Provides
SessionWithInitializedTablesFactory blabla(@Named(InjectionNames.CACHE) KeyspaceConfiguration keyspaceConfiguration,
                                                        InitializedCluster cluster,
                                                        @Named(InjectionNames.CACHE) CassandraModule module) {
        return new SessionWithInitializedTablesFactory(keyspaceConfiguration, cluster.cluster, module);
    }
```

?